test/server/config_validation/xds_verifier.cc

```diff
@@ -0,0 +1,324 @@
+#include "test/server/config_validation/xds_verifier.h"
+
+#include "common/common/logger.h"
+
+namespace Envoy {
+
+XdsVerifier::XdsVerifier(test::server::config_validation::Config::SotwOrDelta sotw_or_delta)
+    : num_warming_(0), num_active_(0), num_draining_(0), num_added_(0), num_modified_(0),
+      num_removed_(0) {
+  if (sotw_or_delta == test::server::config_validation::Config::SOTW) {
+    sotw_or_delta_ = SOTW;
+  } else {
+    sotw_or_delta_ = DELTA;
+  }
+  ENVOY_LOG_MISC(debug, "sotw_or_delta_ = {}", sotw_or_delta_);
+}
+
+/**
+ * get the route referenced by a listener
+ */
+std::string XdsVerifier::getRoute(const envoy::config::listener::v3::Listener& listener) {
+  envoy::config::listener::v3::Filter filter0 = listener.filter_chains()[0].filters()[0];
+  envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager conn_man;
+  filter0.typed_config().UnpackTo(&conn_man);
+  return conn_man.rds().route_config_name();
+}
+
+/**
+ * @return true iff the route listener refers to is in all_routes_
+ */
+bool XdsVerifier::hasRoute(const envoy::config::listener::v3::Listener& listener) {
+  return all_routes_.contains(getRoute(listener));
+}
+
+bool XdsVerifier::hasActiveRoute(const envoy::config::listener::v3::Listener& listener) {
+  return active_routes_.contains(getRoute(listener));
+}
+
+/**
+ * prints the currently stored listeners and their states
+ */
+void XdsVerifier::dumpState() {
+  ENVOY_LOG_MISC(debug, "Listener Dump:");
+  for (const auto& rep : listeners_) {
+    ENVOY_LOG_MISC(debug, "Name: {}, Route {}, State: {}", rep.listener.name(),
+                   getRoute(rep.listener), rep.state);
+  }
+}
+
+/*
+ * if you add a new listener that is not being updated, there are two cases:
+ * 1. the listener has not been added before
+ *    - if envoy knows about the listener's route, it will be added as active, else warming
+ * 1. a listener of the same name has been added before but it now draining
+ *    - if the routes match, the old listener will be added back to active
+ *    - if the routes don't match, the listener will be added as active/warming as relevant
+ *
+ * if a listener is updated (i.e. there is a already a listener by this name), there are 3 cases:
+ * 1. the old listener is active and the new is warming:
+ *    - old will remain active
+ *    - new will be added as warming, to replace the old when it gets its route
+ * 2. the old listener is active and new is active:
+ *    - old's filter chain is drained (seemingly instantaneously)
+ *    - new is added as active
+ * 3. the old listener is warming and new is active/warming:
+ *    - old is completely removed
+ *    - new is added as warming/active as normal
+ */
+
+/**
+ * update a listener when its route is changed, draining/removing the old listener and adding the
+ * updated listener
+ */
+void XdsVerifier::listenerUpdated(envoy::config::listener::v3::Listener listener) {
```

Comment:
I could be wrong about this, but since `listenerUpdated/listenerAdded` are called from `XdsFuzzTest::addListener`, if I pass it by `const auto&` (since I'm adding the listeners to a vector) it will lead to undefined behaviour when `XdsFuzzTest::addListener` returns and listener proto that is built there goes out of scope?

Reply:
Are you capturing listener somewhere where it lives outside of the lexical scope? Or is all use of listener strictly nested? Even if you are capturing, I recommend passing by reference all the way down until the point at which you really want to copy/retain, and then assigning to a non-ref variable. This will make reasoning about performance easier.

Reply:
Fixed! It seems C++ will auto cast the ref to a non-ref when I add it to the vector